### ISSUE TITLE
General Continuous Deployment complaints issue

### ISSUE THREAD

**Issue by DRMacIver** [HypothesisWorks/hypothesis-python#555]
So we now do continuous deployment: Every merge to master that changes src/ will get released to pypi as soon as the build goes green on Travis.

This is a big deal and a big change, and it won't be without its annoyances.

This is a general issue thread for complaining about and discussing those annoyances, both for maintainers and users of Hypothesis. If something about this process bugs you, here's the place to discuss it.  We can then farm off specific issues for ways to improve on or mitigate those annoyances as a result of that discussion.

**Comment by Lukasa** [comment]
So, the biggest source of annoyance for me right now relates to pinned test dependencies.

At this time, all of the python-hyper projects pin to specific versions of all their test dependencies, including to the patch release version. The goal of this is to ensure that all tests always run identical code (or as close to that as possible), and so to isolate problems caused by revving test dependencies from those caused by project-level code changes.

This is managed by having requires.io check when our test dependencies have released, and opening PRs that bump the version. These PRs then run the tests, allowing us to spot when our dependencies have broken us or revealed errors in our code.

As hypothesis is the second-most prevalent test dependency across the python-hyper project (after pytest), this means that each hypothesis release causes a flurry of builds. For example, for 3.7.1 it caused:

- python-hyper/hyper-h2#545
- python-hyper/hyper-h2#546
- python-hyper/hyper-h2#547
- python-hyper/hyper-h2#548
- python-hyper/hyper-h2#549
- python-hyper/hyper-h2#550
- python-hyper/hyper-h2#551
- python-hyper/hyper-h2#552
- python-hyper/priority#63
- python-hyper/brotlipy#75
- python-hyper/hpack#102

This has been moderately annoying because these flurries of builds clog up the Travis queue pretty badly, but it's generally been an acceptable cost of doing business because of the relative infrequency of this event. If we move to expecting this to happen with substantial frequency, then I'll have to go to one of the following models:

1. Pin hypothesis to a minor release, not a patch release. That works great if you never ship patch releases that have bugs, but all due respect I don't trust you (or anyone else) that much. :wink:
2. Further-automate this process such that human intervention is not required as long as the builds pass. This is definitely do-able, but requires me to write a custom github bot that will auto-merge these PRs (and *only* these PRs) as long as the build is green.
3. Something else I guess?

**Comment by DRMacIver** [comment]
One possible suggestion for 3 (which is absolutely intended as a suggestion, not as a "you should definitely do this") . is "Switch to pyup.io and [set your PR frequency](https://pyup.io/docs/configuration/) to something you can manage.": You can configure the freqeuncy here.

**Comment by DRMacIver** [comment]
FWIW, I think pinning Hypothesis to a minor version is explicitly a non-solution *even if we are perfect and never introduce bugs in patch releases*. Patch releases may improve Hypothesis's bug finding capabilities, which may cause your build to start failing even if nothing has changed on your end and Hypothesis has only got better. You probably still want to silo those in PRs.

**Comment by DRMacIver** [comment]
I think, in general, my stance on the overhead of merging updates problem is that I acknowledge it and am sympathetic to it but don't think we should do anything to mitigate it at this time.

My reasoning is roughly this:

* I think the annoyance to heavy users of Hypothesis is probably outweighed by the general benefit of giving the larger community of users access to bug fixes, improvements, new functionality, ASAP
* There is no level of releases at which this *isn't* a problem, and it's almost impossible to find a sensible balance here.
* I think if this really does end up being that annoying in practice then a solution will be found one way or another - either in us scaling back to, say, weekly releases, or in further automation on the part of the people using Hypothesis this way.

@Lukasa I'd encourage you to let me know if you e.g. find this makes you reluctant to add new usage of Hypothesis, or more generally if it really is just intensely annoying in practice.

**Comment by DRMacIver** [comment]
I think my supplemental opinion is "Hurrah! Hypothesis found a bug in your workflow. 😁 😜"

**Comment by DRMacIver** [comment]
Hypothesis has also found a bug in my workflow.

Consider #502. This is a pure refactoring PR. It seems excessively annoying to release a new version for that, even by my standards.

I propose the following refined model:

* master remains the released version, merges of source changes into master still produce releases.
* Create a second long-running protected branch called upcoming. upcoming is for, well, upcoming changes that we have not yet released.
* Suppress the changelog date check when the base branch is upcoming.
* When we want to release what's in upcoming we create a pull request into master based off its current contents.

I still want us to release fairly rapidly, but this at least stops us *completely* wasting @Lukasa's time. ;-)

Does this seem reasonable to people?

**Comment by Lukasa** [comment]
So, this will seem like me being super contrarian but...I think if you believe in this release model then you should also do patch releases for refactorings.

Refactorings are, generally speaking, no more-or-less likely to cause problems than any other release. They can still break people's code, or cause performance improvements, or anything else. If the goal here is to be continually releasing, then frankly I think you should embrace that goal wholeheartedly. If it's better to get code out the door faster, then that should apply to all code IMHO: it avoids the need to distinguish between "refactoring" and "improvement".

Let my problems be my problems. :wink:

**Comment by DRMacIver** [comment]
I think that's a reasonable position and I'm tempted to go with it. I guess I'm just having cold feet.

A consequence of this is that refactorings start to go into the changelog. Does this seem like a good or a bad thing?

**Comment by Lukasa** [comment]
You could adopt the iOS model and just have your changelog read "Minor bug fixes and performance improvements" each release. ;)

More seriously, you can handle it a few different ways, I think. The best one is probably to be up front: "Refactored the `given` function". That helps point users to the things that have changed in case a refactoring actually breaks their code in some way. Other options are to have a generic message, or just to say nothing at all and have a line that reads "No notable changes" or the equivalent. But I think honesty here is best.

**Comment by alexwlchan** [comment]
I agree with @Lukasa – we should just release minor versions for each refactor. Having to keep an "upcoming" or "staging" branch feels like a lot of unnecessary hassle. And "refactored the `given` function" is a perfectly reasonable changelog, and gives me a clue where to look if I suddenly see potential weirdness.

**Comment by DRMacIver** [comment]
OK. Lets stick to our guns for now and do releases for refactorings. Thanks for talking me down, both of you :-)

**Comment by Zac-HD** [comment]
Not asking for changes, but just noting that it's going to be annoying to keep the changelog in order when multiple pulls are being merged (merge conflicts) and to get the date right on a pull that might wait a while for merge (either you update it daily, or have the build keep failing and still have to update before merge).  Plus the Travis build might pass, then the date changes before merge and the deploy build on master fails.

**Comment by DRMacIver** [comment]
> Not asking for changes, but just noting that it's going to be annoying to keep the changelog in order when multiple pulls are being merged (merge conflicts)

Yup, it will. I'd like to figure out a way to make this better.

But note that this is not actually directly caused by the continuous delivery - it's just caused by the implied requirement that we should be writing changelog entries when we write the feature, not when we do the release. I think this is a *really good* requirement and solves a lot of problems (historically I've missed out a lot of things when writing the changelog - e.g. I realised about half an hour ago that #517 got missed out of changelog for 3.7.1 simply because I forgot).

I feel pretty strongly that even if we drop the continuous delivery we should keep the policy of updating the changelog, so I think the solution to this one is going to be to come up with a better changelog workflow.

> and to get the date right on a pull that might wait a while for merge (either you update it daily, or have the build keep failing and still have to update before merge). Plus the Travis build might pass, then the date changes before merge and the deploy build on master fails.

Yeah, this is annoying, but fortunately it fails in the safe direction: We get a broken build on master that doesn't deploy and someone has to take 5 minutes to fix it and issue a new pull request.

If this ends up happening a lot we could look into some better automation of versioning and keeping the changelog up to date.

**Comment by Zac-HD** [comment]
Well, #536 /  3.8.0 has not been released because [the deploy script doesn't think it's on master](https://travis-ci.org/HypothesisWorks/hypothesis-python/builds/224911175).

@DRMacIver I'm going to leave this in your hands.

**Comment by DRMacIver** [comment]
I think this is a race condition in the build! It's getting deployed as part of https://travis-ci.org/HypothesisWorks/hypothesis-python/jobs/224911450 instead.

I guess the correct logic is actually to check if we're in a non pull request build of master on Travis.

**Comment by Zac-HD** [comment]
It would be nice if the changelog entry was copied over to the release on Github.  However I'm not sure this is possible by pushing tags, so we might have to use the releases API.

**Comment by Zac-HD** [comment]
Per the linked issue, conda-forge currently requires someone to manually increment the version for Hypothesis.  Automating this seems like a good idea, and will make me happier at work too.

**Comment by Zac-HD** [comment]
Following #505 / #527, we should also have a "your first pull request" checklist which covers things like writing the changelog entry (should you link to an issue? etc), how CI and CD works (moved from CONTRIBUTING), and so on.  Also an offer to "hand over" incomplete contributions for a maintainer to finish or close.

**Comment by alexwlchan** [comment]
How about having Travis fill in the date of new releases? It’s definitely been a source of frustration (for @Zac-HD in particular) that if you upload a changelog and it doesn't get reviewed for a day or so, you need to bump the release date after the review.

Travis already tags new releases for us – how about having it fill in the release date when it comes to it.

(Although how do you get around master being a protected branch? 🤔 )

**Comment by DRMacIver** [comment]
> Travis already tags new releases for us – how about having it fill in the release date when it comes to it.
> (Although how do you get around master being a protected branch? 🤔 )
 
Yeah I've been wondering along the same lines. I think we clearly want a computer to be doing the annoying reconciliation stuff for versions and changelogs.

For the protected branches thing... I was blocked on that too, but now that I look into it I think it might just work. I'm the one who added the deploy key, so it might count as having administrator access, and now that I check it turns out that my admin powers mean that I've just been able to push straight to master all along *anyway*, I just haven't been doing it.

**Comment by Zac-HD** [comment]
Yes, I would be a fan of automating away release date maintenance ❤️ 

Interestingly, most of `check-changelog.py` would still be useful; we would just test for a token (eg (`date_TBD`) next to the current version, and check that the token only appears once so a simple find/replace works as expected.  Plus maybe that the overline and underline are the right length, since the date won't make it so obvious if the formatting goes wrong.

**Comment by Zac-HD** [comment]
PyPI Error 500:

- [3.8.4](https://travis-ci.org/HypothesisWorks/hypothesis-python/jobs/232734534) not uploaded; needs restart later
- [3.8.5](https://travis-ci.org/HypothesisWorks/hypothesis-python/jobs/232749840) not uploaded; needs restart later

(Note: order of those jobs is critical)

**Comment by Zac-HD** [comment]
I think the last problems in this thread were solved by #787:

- The advice for users is "pick an update schedule you can handle and do that, without reference to our cadence"
- The pain of changelog merging is almost entirely gone
- The race condition is handled by GitHub picking up a merge conflict in `RELEASE.txt`!
